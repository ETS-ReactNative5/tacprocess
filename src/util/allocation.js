import { isFloat } from "../util";


export function illegalAllocation(element, index, array){
  if(
    !(isFloat(element.allocatedTime.p0)) ||
    !(isFloat(element.allocatedTime.p1)) ||
    !(isFloat(element.allocatedTime.p2)) ||
    !(isFloat(element.allocatedTime.p3)) ||
    !(isFloat(element.allocatedTime.p4)) ||
    !(parseFloat(element.allocatedTime.p0) >= 0) ||
    !(parseFloat(element.allocatedTime.p1) >= 0) ||
    !(parseFloat(element.allocatedTime.p2) >= 0) ||
    !(parseFloat(element.allocatedTime.p3) >= 0) ||
    !(parseFloat(element.allocatedTime.p4) >= 0)
){
    return true
  }
  return false
}

export function checkAllocatedTimes(proposals){
  /*
   *
   *
   * @param proposals and array of proposals
   * @return true if all the allocated times correct else false
   */
  return !proposals.some(illegalAllocation)
}

export function getQuaryToAddAllocation(proposals, partner, semester){
  const pList = []
  if (checkAllocatedTimes(proposals)){

    proposals.forEach(p => (
      [0, 1, 2, 3, 4].forEach( t => {
        const priority = `p${t}`
        pList.push(
        `{
            proposalCode: "${p.proposalCode}",
            priority: ${t},
            time: ${p.allocatedTime[priority]}
         }`
      )}
    )
    ))
  }
  const mutateQuery = `
      mutation uta {
        updateTimeAllocations(timeAllocations: {
          partner: "${ partner }",
          semester: "${ semester }",
          timeAllocations: [
            ${pList}
          ]
        }) {
          success
        }
      }
  `
  return mutateQuery
}
