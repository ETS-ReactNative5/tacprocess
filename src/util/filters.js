import { ALL_PARTNER,
	HOME_PAGE,
	STATISTICS_PAGE,
	DOCUMENTATION_PAGE,
	TECHNICAL_PAGE,
	TAC_PAGE,
	ADMIN_PAGE
} from "../types"


export function totalTimeRequestedPerParner(proposals, semester, partner="All" ){
	/*
	*
	* @param proposals of a selected partner all or single
	* @param semester selected semester
	* @param partner all partners or a single
	* @return total
	*/
	
	let total = 0;
	proposals.forEach( p =>{
		p.timeRequests.forEach( r => {
			if (r.semester === semester){
				r.distribution.forEach(d => {
					if ( partner === "All"){
						total += d.time
					}else if (d.partnerCode === partner){
						total += d.time
					}
				})
				
			}
		})
	});
	return total
}

export const semestersArray = () => {
	let startYear = 2006;
	const today = new Date();
	const year = today.getFullYear();
	let semester = [];
	while (startYear < year + 8){
		semester.push(
			`${ startYear }-1`, `${ startYear }-2`
		);
		startYear += 1
	}
	return semester
};

export const firstSelectedPartner = roles => {
	let first = ALL_PARTNER;
	
	for (let r of roles || []) {
		if (r.type === "ADMINISTRATOR") {
			first = ALL_PARTNER;
			break;
		}
		if (r.type === "SALT_ASTRONOMER") {
			first = ALL_PARTNER;
			break;
		}
		if (r.type === "TAC_CHAIR") {
			first = r.partners[0] || "";
		}
		if (r.type === "TAC_MEMBER") {
			first = r.partners[0] || "";
		}
	}
	
	return first
};

export const getPartnerList = roles => {
	let partnerList = [];
	for (let r of roles || []) {
		if (r.type === "ADMINISTRATOR" || r.type === "SALT_ASTRONOMER") {
			partnerList = r.partners;
			partnerList.includes(ALL_PARTNER) ? partnerList.push() : partnerList.push(ALL_PARTNER);
			break;
		}
		
		if (r.type === "TAC_CHAIR") {
			partnerList = r.partners;
		}
		if (r.type === "TAC_MEMBER") {
			partnerList = r.partners;
		}
	}
	return partnerList
};

/*
* @params list an array of
*/
export const listForDropdown = list => {
	return (list || []).map( l => ({ label: l, value: l }))
};

/**
 * method convert a standard SA object to an array of SA names
 *
 * @params Array of SALT astronomers directly from server
 * @return Array of SALT astronomers names
 * */
export const getAstronomersList = saList => {
	return (saList || []).map( l => (`${l.name}`))
};

/**
 * get a current uri path and return the selected page name
 *
 * @params pathname uri path
 * @return name of selected page or Home page by default
 */
export const loadedPage = pathname => {
	console.log({pathname});
	return pathname === "/"? HOME_PAGE :
		pathname === "/timeallocation"? TAC_PAGE :
			pathname === "/statistics"? STATISTICS_PAGE :
				pathname === "/techreview"? TECHNICAL_PAGE :
					pathname === "/documentation"? DOCUMENTATION_PAGE :
						pathname === "/admin"? ADMIN_PAGE : HOME_PAGE;
};

/**\
 * It reduce the proposals to only proposals that are assigned to @param astronomer
 * if @param astronomer is "All" all are returned
 * if @param astronomer is "Assigned" only assigned proposals returned
 * if @param astronomer is "Not Assigned" only unassigned proposals returned
 * if @param astronomer is "SA name" only proposals that are assigned to "SA name" can be returned
 *
 * @param proposals array of proposals
 * @param astronomer a SALT astronomer
 * @return proposals reduced proposals
 * */

export const reduceProposalsPerAstronomer = (proposals, astronomer) => {
	let prop = [];
	if (astronomer === "All"){
		prop = proposals
	}
	else if (astronomer === "Assigned"){
		proposals.forEach(p => {
			if (p.liaisonAstronomer !== null) {prop.push(p)}
		})
	}
	else if (astronomer === "Not Assigned"){
		proposals.forEach(p => {
			if (p.liaisonAstronomer === null) {prop.push(p)}
		})
	}else {
		proposals.forEach(p => {
			if (p.liaisonAstronomer === astronomer) {prop.push(p)}
		})
	}
	
	return prop
};
