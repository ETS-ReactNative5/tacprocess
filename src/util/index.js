import * as types from '../types';

/**
 * Get the observing time for a semester in a proposal.
 * If a partner is given, the observing time is the time requested for that partner.
 *
 * @param proposal
 * @param semester
 * @param partner
 * @returns {*}
 */
export function proposalObservingTime(proposal, semester, partner='All') {
    return proposal.timeRequests
            .filter(r => r.semester === semester) // semester is correct
            .reduce((allDistributionItems, r) => [...allDistributionItems, ...r.distribution], []) // collect all partner time requests
            .filter(d => partner === 'All' || d.partnerCode === partner) // partner is correct
            .reduce((sum, d) => sum + d.time, 0); // add up all time requests
}
/**
 * Get the observing time requested for a semester and transparency in a list of proposals.
 * If a partner is given, the observing time is calculated for that partner only.
 *
 * @param proposals
 * @param semester
 * @param transparency
 * @param partner
 */
export function observingTimeForTransparency(proposals, semester, transparency, partner='All') {
    return proposals
            .filter(p => p.transparency === transparency) // transparency is correct
            .reduce((allTimeRequests, p) => [...allTimeRequests, ...p.timeRequests], []) // collect time requests
            .filter(r => r.semester === semester) // semester is correct
            .reduce((allDistributionItems, r) => [...allDistributionItems, ...r.distribution], []) // collect partner time requests
            .filter(d => partner === 'All' || d.partnerCode === partner) // partner is correct
            .reduce((sum, request) => sum + request.time, 0); // add up all time requests
}

/**
 * Get the observing time requested for a semester and seeing range in a list of proposals.
 * If a partner is given, the observing time is calculated for that partner only.
 *
 * The seeing range must be an array of the minimum and maximum value the proposal's maximum
 * seeing may have. The minimum value is inclusive, the maximum value exclusive.
 *
 * @param proposals
 * @param semester
 * @param seeingRange
 * @param partner
 */
export function observingTimeForSeeing(proposals, semester, seeingRange, partner='All') {
    return proposals
            .filter(p => seeingRange[0] <= p.maxSeeing && p.maxSeeing < seeingRange[1])// seeing is correct
            .reduce((allTimeRequests, p) => [...allTimeRequests, ...p.timeRequests], []) // collect time requests
            .filter(r => r.semester === semester) // semester is correct
            .reduce((allDistributionItems, r) => [...allDistributionItems, ...r.distribution], []) // collect partner time requests
            .filter(d => partner === 'All' || d.partnerCode === partner) // partner is correct
            .reduce((sum, request) => sum + request.time, 0); // add up all time requests
}

/**
 * Get the number of instrument configurations in a proposal.
 *
 * @param proposal
 * @returns {*}
 */
export function instrumentCount(proposal) {
    return Object.keys(proposal.instruments)
            .reduce((sum, key) => sum + proposal.instruments[key].length, 0);
}

/**
 * Get the observing time requested for a semester and instrument in a proposal.
 * If a partner is given, the observing time is calculated for that partner only. If a field is
 * given, in the JavaScript object representing the object this field must have the given
 * value.
 *
 * If a proposal has more than one instrument configuration, the time is split evenly between
 * the configurations.
 *
 * The instrument argument is case-insensitive.
 *
 * The partner, field and value must be passed as an object. For example,
 *
 * observingTimeForInstrument(proposals, '2017-1', 'rss', {partner: 'RSA'})
 * observingTimeForInstrument(proposals, '2017-1', 'rss', {partner: 'RSA', field: 'mode', value: 'Polarimetry'})

 * @param proposal
 * @param semester
 * @param instrument
 * @param field
 * @param value
 * @param partner
 */
export function proposalObservingTimeForInstrument(proposal, semester, instrument, {field, value, partner='All'}) {
    const instrumentModeCount = (proposal.instruments[instrument.toLowerCase()] || [])
            .filter(v => !field || v[field] === value)
            .length;

    const instrumentModeFraction = instrumentModeCount / instrumentCount(proposal);

    const totalObservingTime = proposal.timeRequests
            .filter(r => r.semester === semester) // semester is correct
            .reduce((distributionItems, r) => [...distributionItems, ...r.distribution], []) // collect all partner time requests
            .filter(d => partner === 'All' || d.partnerCode === partner) // partner is correct
            .reduce((sum, d) => sum + d.time, 0); // add all time requests

    return instrumentModeFraction * totalObservingTime;
}

/**
 * Get the observing time requested for a semester and instrument in a list of proposals.
 * If a partner is given, the observing time is calculated for that partner only. If a field is
 * given, in the JavaScript object representing the object this field must have the given
 * value.
 *
 * If a proposal has more than one instrument configuration, the time is split evenly between
 * the configurations.
 *
 * The instrument argument is case-insensitive.
 *
 * The partner, field and value must be passed as an object. For example,
 *
 * observingTimeForInstrument(proposals, '2017-1', 'rss', {partner: 'RSA'})
 * observingTimeForInstrument(proposals, '2017-1', 'rss', {partner: 'RSA', field: 'mode', value: 'Polarimetry'})

 * @param proposals
 * @param semester
 * @param instrument
 * @param field
 * @param value
 * @param partner
 */
export function observingTimeForInstrument(proposals, semester, instrument, {field, value, partner}) {
    return proposals
            .reduce((sum, proposal) =>
                            sum + proposalObservingTimeForInstrument(proposal,
                                                                     semester,
                                                                     instrument,
                                                                     {field, value, partner}), 0);
}

/**
 * Get the sorted list of partners a user should see.
 *
 * A partner is included in this list if it is included in any of the user's roles. If the user is an
 * administrator, the partner 'All' is added to the list as well.
 *
 * @param user
 */
export function partners(user) {
    if (!user.roles) {
        return [];
    }
    const initial = hasRole(user, types.ADMINISTRATOR) || hasRole(user, types.SALT_ASTRONOMER) ? [types.ALL_PARTNER] : [];
    const rolePartners = user.roles // collect partners from all roles
            .reduce((prev, role) => [...(role.partners || []), ...prev], initial);
    const partnerSet = new Set(rolePartners);

    return Array.from(partnerSet).sort();
}

/**
 * Check whether a user has a role for a partner.
 *
 * The partner is ignored for the roles of administrator and SALT Astronomer.
 *
 * The partner "All" is not supported.
 *
 * @param user
 * @param role
 * @param partner
 */
export function hasRole(user, role, partner) {
    if (role === types.ADMINISTRATOR || role === types.SALT_ASTRONOMER) {
        return (user.roles || []).some(r => r.role === role);
    } else {
        return (user.roles || []).some(r => r.role === role && (r.partners || []).includes(partner));
    }
}

export function canDo(user, action, partner) {
    switch (action) {
    case types.VIEW_TIME_ALLOCATION_PAGE:
        return hasRole(user, types.TAC_MEMBER, partner) ||
                hasRole(user, types.SALT_ASTRONOMER) ||
                hasRole(user, types.ADMINISTRATOR);
    case types.EDIT_TIME_ALLOCATION_PAGE:
        return hasRole(user, types.TAC_CHAIR, partner) ||
                hasRole(user, types.ADMINISTRATOR);
    default:
        return false;
    }
}

export function isFloat(val) {
    const floatRegex = /^[+-]?\d+(?:[.,]\d*?)?$/;
    if (!floatRegex.test(val))
        return false;

    const temp = parseFloat(val);
    if (isNaN(temp))
        return false;
    return true;
}

export function canUserWriteAllocations(user, partner){
  let canWrite = false;
  user.partners.forEach( p => {
    if (p.value === partner){
      if (p.role === "ADMINISTRATOR" ||
          p.role === "TAC_CHAIR"){
        canWrite = true;
      }
    }
  });
  console.log("Can Wrte: ", canWrite);
  return canWrite;
}

export function canUserWriteTechComments(user, partner){
  let canWrite = false;
  user.partners.forEach( p => {
    if (p.value === partner){
      if (p.role === "ADMINISTRATOR" ||
          p.role === "SALT_ASTRONOMER"){
        canWrite = true;
      }
    }
  });
  return canWrite;
}
