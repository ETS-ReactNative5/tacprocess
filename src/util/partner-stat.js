/**
 * A function that sums up the numbers in a list or array and returns the total
 * @param array
 * @returns {number}
 */
import { getPercentage, rounded } from './index'

export const sumNumbers = (array) => array.reduce((a, b) => a + b, 0)

/**
 * A function to calculate total observation
 * @param completionStats
 * @returns {number}
 */
export const calculateTotalObservation = (completionStats) => {

  const allCompletionStats = completionStats
  .filter(c => c.partner.toUpperCase() === 'ALL').shift().summary.observedTime

  return allCompletionStats.p0 + allCompletionStats.p1 + allCompletionStats.p2 + allCompletionStats.p3
}

/**
 * A function that calculates the allocated and observed time for the specific proposal with regards to the observation
 * priorities. It also calculates the percentage of those priorities.
 * @param proposal
 * @param priorityType
 * @param statusType
 * @param semester
 * @param partnerCode
 * @returns {
 *  null|
 *  {
 *    percentage: number,
 *    allocatedTime: number,
 *    observedTime: number,
 *    remainder: number
 *  }}
 */
export const statusPriority = (proposal, priorityType, statusType, semester, partnerCode) => {
  if (proposal == null) { return null }
  let timeAllocationPriority = []
  const { observations, timeAllocations } = proposal
  if (partnerCode === 'All') {
    // according to the time allocated for the semester and priority
    timeAllocationPriority = timeAllocations.filter((t) =>
      (t.semester === semester && t.priority === priorityType)
    )
  } else {
    // according to the time allocated for the semester and priority and by the partner
    timeAllocationPriority = timeAllocations.filter((t) =>
      (t.semester === semester && t.partnerCode === partnerCode && t.priority === priorityType)
    )
  }
  // according to the observation status and to the block priority and semester
  const observationsPriority = observations.filter((o) =>
    (o.status === statusType && o.block.priority === priorityType && o.block.semester === semester)
  )

  // calculating the total allocated time
  const totalAllocTime = sumNumbers(timeAllocationPriority.map((t) => t.amount))

  // calculating the total observed time
  const totalObsTime = sumNumbers(observationsPriority.map((o) => o.block.length))
  // calculating the remained allocated time
  const remainingAllocatedTime = totalObsTime ? totalAllocTime - totalObsTime : totalAllocTime

  return {
    priority: priorityType,
    allocatedTime: totalAllocTime,
    observedTime: totalObsTime,
    percentage: rounded(getPercentage(totalObsTime, totalAllocTime)),
    remainder: remainingAllocatedTime
  }
}

/**
 * A function that handles the comment of the Partner's statistics per proposal on a given semester.
 * @param proposal
 * @param semester
 * @returns {string}
 */
export const semesterComment = (proposal, semester) => {
  const { completionComments } = proposal
  if (completionComments.length) {
    return completionComments.filter(c => c.semester === semester).length ? completionComments.filter(c => c.semester === semester)[ 0 ].comment || '' : ''
  }
  return ''
}

/**
 * A function that checks if there has been observation done
 * @param priorityObservations
 * @returns {boolean}
 */
export const neverObserved = (priorityObservations) => priorityObservations.filter(op =>
  typeof (op.percentage) === 'string' ? parseFloat(op.percentage) !== 0 : op.percentage !== 0)

/**
 * A function that calculates the overall completion rate for a proposal.
 * @param proposal
 * @param statusType
 * @param semester
 * @param partnerCode
 * @returns {string|null}
 */
export const overallCompletionRate = (proposal, statusType, semester, partnerCode) => {
  if (proposal == null) { return null }
  let { observations, timeAllocations } = proposal

  if (partnerCode === 'All') {
    // according to the time allocated for the semester
    timeAllocations = timeAllocations.filter(t => (t.semester === semester))
  } else {
    // according to the time allocated for the semester and by the partner
    timeAllocations = timeAllocations.filter(t => (t.semester === semester && t.partnerCode === partnerCode))
  }

  // according to the observation status and semester
  observations = observations.filter(o => (o.status === statusType && o.block.semester === semester))

  // calculating the allocated time according to the priority.
  const allocatedTimeP0 = sumNumbers(timeAllocations.filter(t => t.priority === 0).map((t) => t.amount))
  const allocatedTimeP1 = sumNumbers(timeAllocations.filter(t => t.priority === 1).map((t) => t.amount))
  const allocatedTimeP2 = sumNumbers(timeAllocations.filter(t => t.priority === 2).map((t) => t.amount))
  const allocatedTimeP3 = sumNumbers(timeAllocations.filter(t => t.priority === 3).map((t) => t.amount))

  // calculating the observed time according to the priority.
  const observedTimeP0 = sumNumbers(observations.filter(o => o.block.priority === 0).map((o) => o.block.length))
  const observedTimeP1 = sumNumbers(observations.filter(o => o.block.priority === 1).map((o) => o.block.length))
  const observedTimeP2 = sumNumbers(observations.filter(o => o.block.priority === 2).map((o) => o.block.length))
  const observedTimeP3 = sumNumbers(observations.filter(o => o.block.priority === 3).map((o) => o.block.length))

  // calculating the total allocated time
  const allocatedTime = allocatedTimeP0 + allocatedTimeP1 + allocatedTimeP2 + allocatedTimeP3

  // calculating the total observed time
  const observedTime = (observedTimeP0 + observedTimeP1 + observedTimeP2 + observedTimeP3)

  // completion overall rate
  const completionOverallRate = allocatedTime !== 0
    ? observedTime / (allocatedTimeP0 + allocatedTimeP1 + allocatedTimeP2 + allocatedTimeP3 / 3) : 0

  return allocatedTime !== 0 ? rounded(completionOverallRate) : 'Nan'
}

/**
 * A function for calculating the observed time per partner.
 * All the quantities are for one proposal.
 * @param timeAllocations
 * @param observations
 * @param partnerTimeAllocations
 * @returns {number}
 */
const calculateObservedTime = (timeAllocations, observations, partnerTimeAllocations) => {
  const partnerAllocatedTime = sumNumbers(partnerTimeAllocations.map((t) => t.amount))
  const allocatedTime = sumNumbers(timeAllocations.map((t) => t.amount))
  const observedTime = sumNumbers(observations.map((o) => o.block.length))

  if (!allocatedTime) {
    return 0
  }

  return (partnerAllocatedTime / allocatedTime) * observedTime
}

/**
 * A function for calculating the partner's summary statistics for a specific semester.
 * @param proposals
 * @param semester
 * @param partnerCode
 * @param partnerShareTimes
 * @param totalObservation
 * @returns {
 * {
 *    partnerAllocatedShareTime: number,
 *    partnerObservedShareTime: number,
 *    completeness: {p0p1: (string|number), p2: (string|number), total: (string|number), p3: (string|number)},
 *    allocatedTime: {p0p1: number, p2: number, total: number, p3: number},
 *    observedTime: {p0p1: number, p2: number, total: number, p3: number}
 * }|null
 * }
 */
export const partnerSummaryStat = (proposals, semester, partnerCode, partnerShareTimes, totalObservation) => {
  if (!proposals.length) { return null }

  let partnerAllocatedShareTime = 0

  if (partnerShareTimes && partnerShareTimes.length) {
    // if no partner is selected, sum the share times of all the partner
    partnerAllocatedShareTime = partnerCode === 'All'
      ? sumNumbers(partnerShareTimes.map((partnerShareTime) => partnerShareTime.sharePercent))
      : partnerShareTimes.find((partnerShareTime) => partnerShareTime.partnerCode === partnerCode).sharePercent
  }

  let p0p1Allocated = 0
  let p2Allocated = 0
  let p3Allocated = 0

  let p0p1Observed = 0
  let p2Observed = 0
  let p3Observed = 0
  proposals.forEach(p => {
    // filter the time allocated by semester and priorities
    const p0p1TimeAllocations = p.timeAllocations.filter((t) =>
      (t.semester === semester && (t.priority === 0 || t.priority === 1))
    )

    const p2TimeAllocations = p.timeAllocations.filter((t) =>
      (t.semester === semester && t.priority === 2)
    )

    const p3TimeAllocations = p.timeAllocations.filter((t) =>
      (t.semester === semester && t.priority === 3)
    )

    // filter the time allocated by partner
    // if no partner is selected, all the partners are included
    const p0p1PartnerTimeAllocations = partnerCode === 'All' ? p0p1TimeAllocations
      : p0p1TimeAllocations.filter((t) => t.partnerCode === partnerCode)
    const p2PartnerTimeAllocations = partnerCode === 'All' ? p2TimeAllocations
      : p2TimeAllocations.filter((t) => t.partnerCode === partnerCode)
    const p3PartnerTimeAllocations = partnerCode === 'All' ? p3TimeAllocations
      : p3TimeAllocations.filter((t) => t.partnerCode === partnerCode)

    // filter the observed time by status, semester and priorities
    const p0p1Observations = p.observations.filter((o) =>
      (o.status === 'ACCEPTED' && o.block.semester === semester && (o.block.priority === 0 || o.block.priority === 1))
    )

    const p2Observations = p.observations.filter((o) =>
      (o.status === 'ACCEPTED' && o.block.semester === semester && o.block.priority === 2)
    )

    const p3Observations = p.observations.filter((o) =>
      (o.status === 'ACCEPTED' && o.block.semester === semester && o.block.priority === 3)
    )

    // calculate the time allocated per proposal according to the priorities (P0+P1, P2, P3)
    p0p1Allocated += sumNumbers(p0p1PartnerTimeAllocations.map((t) => t.amount))
    p2Allocated += sumNumbers(p2PartnerTimeAllocations.map((t) => t.amount))
    p3Allocated += sumNumbers(p3PartnerTimeAllocations.map((t) => t.amount))

    // calculate the observed time per proposal according to priorities and partner or all partners
    p0p1Observed += calculateObservedTime(p0p1TimeAllocations, p0p1Observations, p0p1PartnerTimeAllocations)
    p2Observed += calculateObservedTime(p2TimeAllocations, p2Observations, p2PartnerTimeAllocations)
    p3Observed += calculateObservedTime(p3TimeAllocations, p3Observations, p3PartnerTimeAllocations)
  })

  // calculating the total allocated time for the partner or all partners
  const totalAllocated = p0p1Allocated + p2Allocated + p3Allocated
  // calculating the total observed time for the partner or all the partners
  const totalObserved = p0p1Observed + p2Observed + p3Observed
  // Observed share time in percent
  const partnerObservedShareTime = getPercentage(totalObserved, totalObservation)

  return {
    allocatedTime: {
      p0p1: p0p1Allocated,
      p2: p2Allocated,
      p3: p3Allocated,
      total: totalAllocated
    },
    observedTime: {
      p0p1: rounded(p0p1Observed, 0),
      p2: rounded(p2Observed, 0),
      p3: rounded(p3Observed, 0),
      total: rounded(totalObserved, 0)
    },
    completeness: {
      p0p1: rounded(getPercentage(p0p1Observed, p0p1Allocated)),
      p2: rounded(getPercentage(p2Observed, p2Allocated)),
      p3: rounded(getPercentage(p3Observed, (p3Allocated / 3))),
      total: rounded(getPercentage(totalObserved, p0p1Allocated + p2Allocated + p3Allocated / 3))
    },
    partnerAllocatedShareTime,
    partnerObservedShareTime: rounded(partnerObservedShareTime)
  }
}
