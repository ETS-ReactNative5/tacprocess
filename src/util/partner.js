import _ from "lodash"


export function totalTimeRequestedPerParner(proposals, semester, partner="All" ){
  /*
  *
  * @param proposals of a selected partner all or single
  * @param semester selected semester
  * @param partner all partners or a single
  * @return total
  */

  let total = 0
  proposals.forEach( p =>{
    p.timeRequests.forEach( r => {
      if (r.semester === semester){
        r.distribution.forEach(d => {
          if ( partner === "All"){
            total += d.time
          }else if (d.partnerCode === partner){
              total += d.time
            }
        })

      }
    })
  })
  return total
}

export function userPartners(user) {
  let allPartners = []
  if (! _.isEmpty(user)){
    user.user.role.forEach(u => {
        u.partners.forEach( s => {
          if ( _.findIndex(allPartners, { value: s, label: s }) === -1 ){
            allPartners.push({ value: s, label: s })
          }
        }
      );
    })
  }
  allPartners = [{value: "All", label: "All"}].concat(allPartners)
  return allPartners
}
