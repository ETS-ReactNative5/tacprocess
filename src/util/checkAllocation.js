import { isFloat } from "../util";


export function illegalAllocation(element, index, array){
  if(
    !(isFloat(element.allocatedTime.p0)) ||
    !(isFloat(element.allocatedTime.p1)) ||
    !(isFloat(element.allocatedTime.p2)) ||
    !(isFloat(element.allocatedTime.p3)) ||
    !(isFloat(element.allocatedTime.p4)) ||
    !(parseFloat(element.allocatedTime.p0) >= 0) ||
    !(parseFloat(element.allocatedTime.p1) >= 0) ||
    !(parseFloat(element.allocatedTime.p2) >= 0) ||
    !(parseFloat(element.allocatedTime.p3) >= 0) ||
    !(parseFloat(element.allocatedTime.p4) >= 0)
){
    return true
  }
  return false
}

export default function checkAllocatedTimes(proposals){
  /*
   *
   *
   * @param proposals and array of proposals
   * @return true if all the allocated times correct else false
   */
  return !proposals.some(illegalAllocation)
}
