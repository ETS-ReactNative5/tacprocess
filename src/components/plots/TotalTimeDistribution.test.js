import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount, render } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import TotalTimeDistribution from "./TotalTimeDistribution";
import {proposalObservingTime} from '../../util';

const proposals = [
  {
    transparency: 'Clear',
    maxSeeing: 2,
    timeRequests: [
      {
        semester: '2018-1',
        distribution: [ { partnerCode: 'RSA', time: 150 }, { partnerCode: 'UW', time: 200 }, { partnerCode: 'IUCAA', time: 250 } ]
      },
      {
        semester: '2018-1',
        distribution: [ { partnerCode: 'RSA', time: 400 }, { partnerCode: 'UW', time: 50 } ]
      }
    ]
  },
  {
      transparency: 'Thin cloud',
      maxSeeing: 2,
      timeRequests: [
        {
          semester: '2017-1',
          distribution: [ { partnerCode: 'RSA', time: 150 }, { partnerCode: 'UW', time: 200 }, { partnerCode: 'IUCAA', time: 250 } ]
        },
        {
          semester: '2018-1',
          distribution: [ { partnerCode: 'RSA', time: 400 }, { partnerCode: 'UW', time: 50 } ]
        }
      ]
    }
];

// Checking if the TotalTimeDistribution Component renders correctly for different input
describe("TotalTimeDistribution Component", () => {
  const semester = '2017-1';
  const partners = ['UW', 'RSA'];

  //Enzyme method testing if it renders correctly with empty values of the proposals, semester, and partner
  it("Render the TotalTimeDistribution Component having unpopulated props with no errors", () => {
    const rendered = mount(<TotalTimeDistribution proposals={[]} semester={""} partner={""}  />);
    expect(shallowToJson(rendered)).toMatchSnapshot();
  });

  it("Render the TotalTimeDistribution Component having populated props with no errors", () => {
    const rendered = mount(<TotalTimeDistribution proposals={proposals} semester={semester} partner={partners[0]}  />);
    expect(shallowToJson(rendered)).toMatchSnapshot();
  });

  it("Render the TotalTimeDistribution Component having populated props with no errors", () => {
    const rendered = mount(<TotalTimeDistribution proposals={proposals} semester={semester} partner={partners[1]}  />);
    expect(shallowToJson(rendered)).toMatchSnapshot();
  });
});

// Checking the observing time for the TotalTimeDistribution
describe("Observing time for the TotalTimeDistribution", () => {
  //Testing for a wrong semester
  it('Should return zero for a wrong semester', () => {
    const result = proposalObservingTime(proposals[0], '2019-1');
    expect(result).toBe(0);
  });
  it('Should return zero for a wrong semester', () => {
    const result = proposalObservingTime(proposals[1], '2019-1');
    expect(result).toBe(0);
  });
  //Testing for the correct semester and partner
  it("Should return the correct observing time for the TotalTimeDistribution for a semester and partner", () => {
    const result = proposalObservingTime(proposals[0], '2018-1', 'RSA');
    expect(result).toBe(550);
  });
  it("Should return the correct observing time for the TotalTimeDistribution for a semester and partner", () => {
    const result = proposalObservingTime(proposals[1], '2017-1', 'RSA');
    expect(result).toBe(150);
  });
  //Testing for the correct semester and all partners
  it("Should return the correct observing time for the TotalTimeDistribution for a semester and all partner", () => {
    const result = proposalObservingTime(proposals[0], '2018-1', 'All');
    expect(result).toBe(1050);
    const result1 = proposalObservingTime(proposals[0], '2018-1');
    expect(result1).toBe(1050);
  });
  it("Should return the correct observing time for the TotalTimeDistribution for a semester and all partner", () => {
    const result = proposalObservingTime(proposals[1], '2017-1', 'All');
    expect(result).toBe(600);
    const result1 = proposalObservingTime(proposals[1], '2017-1');
    expect(result1).toBe(600);
  });
});
