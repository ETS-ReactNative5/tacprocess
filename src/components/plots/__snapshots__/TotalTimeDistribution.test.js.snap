// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotalTimeDistribution Component Render the TotalTimeDistribution Component having populated props with no errors 1`] = `
<TotalTimeDistribution
  partner="UW"
  proposals={
    Array [
      Object {
        "maxSeeing": 2,
        "timeRequests": Array [
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 150,
              },
              Object {
                "partnerCode": "UW",
                "time": 200,
              },
              Object {
                "partnerCode": "IUCAA",
                "time": 250,
              },
            ],
            "semester": "2018-1",
          },
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 400,
              },
              Object {
                "partnerCode": "UW",
                "time": 50,
              },
            ],
            "semester": "2018-1",
          },
        ],
        "transparency": "Clear",
      },
      Object {
        "maxSeeing": 2,
        "timeRequests": Array [
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 150,
              },
              Object {
                "partnerCode": "UW",
                "time": 200,
              },
              Object {
                "partnerCode": "IUCAA",
                "time": 250,
              },
            ],
            "semester": "2017-1",
          },
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 400,
              },
              Object {
                "partnerCode": "UW",
                "time": 50,
              },
            ],
            "semester": "2018-1",
          },
        ],
        "transparency": "Thin cloud",
      },
    ]
  }
  semester="2017-1"
>
  <div
    className="stat-item"
  >
    <svg
      className="plot"
      height={700}
      width={700}
    />
  </div>

</TotalTimeDistribution>
`;

exports[`TotalTimeDistribution Component Render the TotalTimeDistribution Component having populated props with no errors 2`] = `
<TotalTimeDistribution
  partner="RSA"
  proposals={
    Array [
      Object {
        "maxSeeing": 2,
        "timeRequests": Array [
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 150,
              },
              Object {
                "partnerCode": "UW",
                "time": 200,
              },
              Object {
                "partnerCode": "IUCAA",
                "time": 250,
              },
            ],
            "semester": "2018-1",
          },
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 400,
              },
              Object {
                "partnerCode": "UW",
                "time": 50,
              },
            ],
            "semester": "2018-1",
          },
        ],
        "transparency": "Clear",
      },
      Object {
        "maxSeeing": 2,
        "timeRequests": Array [
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 150,
              },
              Object {
                "partnerCode": "UW",
                "time": 200,
              },
              Object {
                "partnerCode": "IUCAA",
                "time": 250,
              },
            ],
            "semester": "2017-1",
          },
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 400,
              },
              Object {
                "partnerCode": "UW",
                "time": 50,
              },
            ],
            "semester": "2018-1",
          },
        ],
        "transparency": "Thin cloud",
      },
    ]
  }
  semester="2017-1"
>
  <div
    className="stat-item"
  >
    <svg
      className="plot"
      height={700}
      width={700}
    />
  </div>

</TotalTimeDistribution>
`;

exports[`TotalTimeDistribution Component Render the TotalTimeDistribution Component having unpopulated props with no errors 1`] = `
<TotalTimeDistribution
  partner=""
  proposals={Array []}
  semester=""
>
  <div
    className="stat-item"
  >
    <svg
      className="plot"
      height={700}
      width={700}
    />
  </div>

</TotalTimeDistribution>
`;
